File: solutions/innerprod_pad.c
/* innerprod.c */
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>

#define N 10000000
#define PAD 16

int main(void)
{
    int *a, *b, ip, i, (*localips)[PAD];
    double t1, t2;

    a = malloc(N * sizeof(*a));
    b = malloc(N * sizeof(*b));
    localips = malloc(omp_get_max_threads()*sizeof(*localips));

#pragma omp parallel for default(none) private(i) shared (a,b)
    for (i = 0; i < N; i++) {
        a[i] = 2;
        b[i] = 3;
    }

    t1 = omp_get_wtime();
    ip = 0;

#pragma omp parallel default(none) private(i) shared (ip,a,b,localips)
    {
      int tid = omp_get_thread_num();
      localips[tid][0] = 0;
#pragma omp for
      for (i = 0; i < N; i++)
	localips[tid][0] += a[i] * b[i];
#pragma omp single
      for (i = 0; i < omp_get_num_threads(); i++)
	ip += localips[i][0];
    }

    t2 = omp_get_wtime();

    printf("Inner product = %d, time=%g\n", ip, t2-t1);
    return 0;
}

File: solutions/innerprod_pad.f90
PROGRAM innerprod

USE omp_lib
IMPLICIT NONE

INTEGER, PARAMETER :: N = 10000000, PAD=16
INTEGER, ALLOCATABLE :: A(:), B(:), localips(:,:)
INTEGER :: ip, i, tid
DOUBLE PRECISION :: t1, t2

    ALLOCATE(A(N),B(N),localips(PAD,0:OMP_GET_MAX_THREADS()-1))
!$OMP PARALLEL DO DEFAULT(none) PRIVATE(i) SHARED (A,B)
    DO i = 1, N
       A(i) = 2
       B(i) = 3
    ENDDO
!OMP END PARALLEL DO

    t1 = OMP_GET_WTIME()
    ip = 0
!$OMP PARALLEL DEFAULT(none) PRIVATE(i,tid) SHARED (ip,A,B,localips)
    tid = OMP_GET_THREAD_NUM()
    localips(1,tid) = 0
!$OMP DO
    DO i = 1, N
       localips(1,tid) = localips(1,tid) + A(i) * B(i)
    ENDDO
!$OMP END DO
!$OMP SINGLE
    DO i = 0, OMP_GET_NUM_THREADS()-1
       ip = ip + localips(1,i)
    ENDDO
!$OMP END SINGLE
!$OMP END PARALLEL
    t2 = OMP_GET_WTIME()

    WRITE(*,*) 'Inner product = ', ip, 'Time =', t2 - t1

END PROGRAM innerprod
